op-supervisor/supervisor/backend/cross/cycle.go

```
@@ -0,0 +1,303 @@
+			}
+		}
+
+		// Store log count and in-timestamp executing messages
+		logCounts[hazardChainIndex] = logCount
+
+		if len(msgs) > 0 {
+			if _, exists := execMsgs[hazardChainIndex]; !exists {
+				execMsgs[hazardChainIndex] = make(map[uint32]*types.ExecutingMessage)
+			}
+		}
+		for logIdx, msg := range msgs {
+			if msg.Timestamp == inTimestamp {
+				execMsgs[hazardChainIndex][logIdx] = msg
+			}
+		}
+	}
+
+	return logCounts, execMsgs, nil
+}
+
+// buildGraph constructs a dependency graph from the hazard blocks.
+func buildGraph(d CycleCheckDeps, inTimestamp uint64, hazards map[types.ChainIndex]types.BlockSeal) (*graph, error) {
+	g := &graph{
+		inDegree0:     make(map[node]struct{}),
+		inDegreeNon0:  make(map[node]uint32),
+		outgoingEdges: make(map[node][]node),
+	}
+
+	logCounts, execMsgs, err := gatherLogs(d, inTimestamp, hazards)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add nodes for each log in the block, and add edges between sequential logs
+	for hazardChainIndex, logCount := range logCounts {
+		for i := uint32(0); i < logCount; i++ {
+			k := node{
+				chainIndex: hazardChainIndex,
+				logIndex:   i,
+			}
+
+			if i == 0 {
+				// First log in block has no dependencies
+				g.inDegree0[k] = struct{}{}
+			} else {
+				// Add edge: prev log <> current log
+				prevKey := node{
+					chainIndex: hazardChainIndex,
+					logIndex:   i - 1,
+				}
+				g.addEdge(prevKey, k)
+			}
+		}
+	}
+
+	// Add edges for executing messages to their initiating messages
+	for hazardChainIndex, msgs := range execMsgs {
+		for execLogIdx, m := range msgs {
+			// Error if the chain is unknown
+			if _, ok := hazards[m.Chain]; !ok {
+				return nil, ErrExecMsgUnknownChain
+			}
+
+			// Check if we care about the init message
+			initChainMsgs, ok := execMsgs[m.Chain]
+			if !ok {
+				continue
+			}
+			if _, ok := initChainMsgs[m.LogIdx]; !ok {
+				continue
+			}
```

This should return an error (wrapping the `types.ErrConflict`) instead. If the executing message points to a non-existent log then there's no point in resolving any remaining cycles, as the block will be invalid.

---

This check is actually just checking that the init message is within the right timestamp. I've added an additional check to ensure the init message exists.